verification/src/solidity/compiler_fetcher.rs

```diff
@@ -0,0 +1,330 @@
+use self::deserialized_types::{DeserializedCompilerInfo, DownloadPath, ListJson};
+use crate::compiler::{fetcher::Fetcher, version::CompilerVersion};
+use async_trait::async_trait;
+use primitive_types::H256;
+use serde::Deserialize;
+use std::{
+    collections::HashMap,
+    fs::{File, OpenOptions},
+    io::ErrorKind,
+    os::unix::prelude::OpenOptionsExt,
+    path::{Path, PathBuf},
+};
+use thiserror::Error;
+use url::Url;
+
+#[derive(Error, Debug)]
+pub enum ListError {
+    #[error("fetching list json returned error: {0}")]
+    ListJsonFetch(reqwest::Error),
+    #[error("cannot parse list json file: {0}")]
+    ParseListJson(reqwest::Error),
+    #[error("error parsing 'path' field: {0}")]
+    Path(url::ParseError),
+}
+
+#[derive(Debug)]
+pub struct CompilerInfo {
+    pub url: Url,
+    pub sha256: H256,
+}
+
+impl TryFrom<(DeserializedCompilerInfo, &Url)> for CompilerInfo {
+    type Error = url::ParseError;
+
+    fn try_from(
+        (compiler_info, download_url): (DeserializedCompilerInfo, &Url),
+    ) -> Result<Self, Self::Error> {
+        let url = match compiler_info.path {
+            DownloadPath::Url(url) => url,
+            // download_url ends with `.../list.json` but join() will replace this with `filename`
+            DownloadPath::Filename(filename) => download_url.join(&filename)?,
+        };
+        Ok(Self {
+            url,
+            sha256: compiler_info.sha256,
+        })
+    }
+}
+
+#[derive(Default)]
+pub struct CompilerFetcher {
+    releases: HashMap<CompilerVersion, CompilerInfo>,
+    folder: PathBuf,
+}
+
+impl CompilerFetcher {
+    pub async fn new(compilers_list_url: &Url, folder: PathBuf) -> Result<Self, ListError> {
```

Comment:
let's may be make this type abstract about the list by implementing the new like
```rust
pub fn new(releases: HashMap..., folder: PathBuf) -> Self
```
and we could move `list_releases` from impl CompilerFetcher to just function and probably rename it to `fetch_compiler_list`